### PR TITLE
Strip down PeerMap

### PR DESCRIPTION
**Pull request by octol**
Strip down peer_map to the minimum we need it to do.

### CODE REVIEW

File: src/chain/chain.rs [hunk 1]
```diff
@@ -694,6 +695,11 @@ impl Chain {
             .flat_map(EldersInfo::member_nodes)
     }
 
+    pub fn get_p2p_node_from_addr(&self, socket_addr: &SocketAddr) -> Option<&P2pNode> {
```
Comment (jeanphilippeD) [inline src/chain/chain.rs]: can you name it 'find_...` get suggest fast access.

File: src/node.rs [hunk 2]
```diff
@@ -407,6 +407,11 @@ impl Node {
             .map(P2pNode::public_id)
     }
 
+    /// Returns a set of elders we should be connected to.
```
Comment (jeanphilippeD) [inline src/node.rs]: `/// Returns a set of elders we can communicate with`

At the end we won't necessarily be connected, so probably best to not add a comment that will soon be misleading.

File: src/peer_map.rs [hunk 3]
```diff
@@ -33,115 +22,33 @@ impl PeerMap {
...
-            if entry.get().is_empty() {
-                let _ = entry.remove();
-                return Some(conn_info);
-            }
-        }
-
-        None
+    // Marks the connection as severed at the network layer.
+    pub fn disconnect(&mut self, socket_addr: SocketAddr) -> Option<ConnectionInfo> {
+        self.connections.remove(&socket_addr)
```
Comment (jeanphilippeD) [inline src/peer_map.rs]: I like these short functions :-)

File: src/states/adult.rs [hunk 4]
```diff
@@ -530,7 +531,16 @@ impl Approved for Adult {
         let _ = self.chain.poll_relocation();
 
         self.remove_elder(pub_id, outbox)?;
-        self.disconnect(&pub_id);
+
+        if let Some(p2p_node) = self.chain.get_p2p_node(pub_id.name()) {
+            let p2p_node = p2p_node.clone();
+            self.disconnect(&p2p_node.peer_addr());
+        } else {
+            info!(
```
Comment (jeanphilippeD) [inline src/states/adult.rs]: this is a log_or_panic, that should never happen.

File: src/states/adult.rs [hunk 5]
```diff
@@ -589,7 +599,16 @@ impl Approved for Adult {
         info!("{} - handle Relocate: {:?}.", self, details);
         self.chain.remove_member(&details.pub_id);
         self.remove_elder(details.pub_id, outbox)?;
-        self.disconnect(&details.pub_id);
+
+        if let Some(p2p_node) = self.chain.get_p2p_node(details.pub_id.name()) {
+            let p2p_node = p2p_node.clone();
+            self.disconnect(&p2p_node.peer_addr());
+        } else {
+            info!(
```
Comment (jeanphilippeD) [inline src/states/adult.rs]: Same, log_or_panic

File: src/states/bootstrapping_peer.rs [hunk 6]
```diff
@@ -137,7 +137,7 @@ impl BootstrappingPeer {
         };
 
         self.send_direct_message(&dst, DirectMessage::BootstrapRequest(destination));
-        self.peer_map_mut().connect(dst);
+        self.peer_map_mut().connect(dst.peer_addr, dst);
```
Comment (jeanphilippeD) [inline src/states/bootstrapping_peer.rs]: might as well only have a single param and do that within the `connect` function?

File: src/states/common/base.rs [hunk 7]
```diff
@@ -241,19 +245,8 @@ pub trait Base: Display {
...
-            let other_transition = self.handle_peer_lost(pub_id, outbox);
-
-            if let Transition::Stay = transition {
-                transition = other_transition
-            }
-        }
-
-        transition
+        let _ = self.peer_map_mut().disconnect(peer_addr);
+        self.handle_peer_lost(peer_addr, outbox)
```
Comment (jeanphilippeD) [inline src/states/common/base.rs]: That is interesting, why this change?

File: src/states/elder/mod.rs [hunk 8]
```diff
@@ -1059,11 +1060,23 @@ impl Elder {
...
+                .unwrap_or(false)
+        };
         if self
             .peer_map()
-            .connected_ids()
-            .filter(|id| self.chain.our_id() != *id)
-            .any(|id| self.chain.is_peer_our_elder(id))
+            .connected_peers()
+            .filter(|conn_info| &our_connection_info != *conn_info)
+            .any(|conn_info| connection_info_belongs_to_our_elder(conn_info))
```
Comment (jeanphilippeD) [inline src/states/elder/mod.rs]: Seem a bit more complicated than I would expect.

```
       let our_name = self.chain.our_name();
        if self
            .chain
            .our_info()
            .member_map()
            .iter()
            .filter(|(name, _)| name != our_name)
            .any(|(_, p2p_node)| !self.peer_map().has(p2p_node.peer_addr()))  {
            return true;
        } ...

```

File: src/states/elder/mod.rs [hunk 9]
```diff
@@ -1538,7 +1558,15 @@ impl Approved for Elder {
         }
 
         self.remove_elder(pub_id, outbox)?;
-        self.disconnect(&pub_id);
+        if let Some(p2p_node) = self.chain.get_p2p_node(pub_id.name()) {
+            let p2p_node = p2p_node.clone();
+            self.disconnect(&p2p_node.peer_addr());
+        } else {
+            info!(
```
Comment (jeanphilippeD) [inline src/states/elder/mod.rs]: log or panic

File: src/states/elder/mod.rs [hunk 10]
```diff
@@ -1702,7 +1730,15 @@ impl Approved for Elder {
 
         self.chain.remove_member(&pub_id);
         self.remove_elder(pub_id, outbox)?;
-        self.disconnect(&pub_id);
+        if let Some(p2p_node) = self.chain.get_p2p_node(pub_id.name()) {
+            let p2p_node = p2p_node.clone();
+            self.disconnect(&p2p_node.peer_addr());
+        } else {
+            info!(
```
Comment (jeanphilippeD) [inline src/states/elder/mod.rs]: log or paninc?

File: src/chain/chain.rs (hunk 1 above)
Comment (octol) [reply src/chain/chain.rs]: Yes good point!

File: src/states/bootstrapping_peer.rs (hunk 6 above)
Comment (octol) [reply src/states/bootstrapping_peer.rs]: Yep you're right, that would be simpler (and more correct)

File: src/states/elder/mod.rs (hunk 8 above)
Comment (octol) [reply src/states/elder/mod.rs]: Good point, I'll update!

File: src/states/bootstrapping_peer.rs (hunk 6 above)
Comment (octol) [reply src/states/bootstrapping_peer.rs]: We don't have the 1-to-many mapping anymore as we did when the key was `PublicId`

File: src/states/common/base.rs (hunk 7 above)
Comment (octol) [reply src/states/common/base.rs]: We don't have the 1-to-many mapping as when the key was `PublicId`

Comment (jeanphilippeD) [reply src/states/common/base.rs]: Could you be a bit more verbose, I still don't understand why it is ok to call handle_peer_lost here.

Comment (octol) [reply src/states/common/base.rs]: Hmm not sure I understand, we called it before as well?

Comment (jeanphilippeD) [reply src/states/common/base.rs]: Good point, I just completely misread the original loop.

File: src/states/common/approved.rs [hunk 11]
```diff
@@ -413,13 +415,26 @@ pub trait Approved: Base {
...
+        self.peer_map_mut().connect(their_conn_info.clone());
         self.send_direct_message(&their_conn_info, DirectMessage::ConnectionResponse);
 
         Ok(())
     }
 
+    fn disconnect_by_id_lookup(&mut self, pub_id: &PublicId) {
+        if let Some(node) = self.chain().get_p2p_node(pub_id.name()) {
+            let node = node.clone();
+            self.disconnect(&node.peer_addr());
```
Comment (jeanphilippeD) [inline src/states/common/approved.rs]: clone or copy the peer_addr instead of the full node since that is what you use.

File: src/states/elder/mod.rs [hunk 12]
```diff
@@ -629,14 +629,14 @@ impl Elder {
...
-        if !self.peer_map().has(&pub_id) {
+        if !self.peer_map().has(p2p_node.peer_addr()) {
             trace!(
                 "{} - Not yet connected to {} - use p2p_node.",
                 self,
                 p2p_node
             );
             self.peer_map_mut()
-                .insert(pub_id, p2p_node.connection_info().clone());
+                .connect(p2p_node.connection_info().clone());
```
Comment (jeanphilippeD) [inline src/states/elder/mod.rs]: OK, this might be ok for now.
Ideally I'd think these would be gone (i.e only update the peer map on quic_p2p event or disconnect call), does it fail if we remove the following?
I'm thinking we should get a `ConnectedTo` event from quic_p2p when sending that?

```
             self.peer_map_mut()
                .connect(p2p_node.connection_info().clone()); 
```

Same apply for all `connect` calls before a `send_direct_message`

File: src/states/joining_peer.rs [hunk 13]
```diff
@@ -147,12 +147,14 @@ impl JoiningPeer {
             RoutingMessage {
                 content:
                     MessageContent::ConnectionRequest {
-                        conn_info, pub_id, ..
+                        conn_info,
+                        pub_id: _pub_id,
```
Comment (jeanphilippeD) [inline src/states/joining_peer.rs]: you can just remove the pub_id field.

File: src/peer_map.rs [hunk 14]
```diff
@@ -6,24 +6,13 @@
...
-/// 2. Otherwise its a two step process: first, when the connection to the peer is established at
-///    the network layer, call `connect`. Then when their public id is received, call `identify`.
-///    This happens when the peer connects to us and then sends us a message which contains their
-///    public id.
 #[derive(Default)]
 pub struct PeerMap {
-    forward: HashMap<XorName, ConnectionInfo>,
-    reverse: HashMap<SocketAddr, HashSet<PublicId>>,
-    pending: HashMap<SocketAddr, PendingConnection>,
+    connections: HashMap<SocketAddr, ConnectionInfo>,
```
Comment (madadam) [inline src/peer_map.rs]: Now the socket address is stored twice - in the key and in the value. Could we change this to `HashMap<SocketAddr, Certificate>` ?

File: src/states/common/approved.rs [hunk 15]
```diff
@@ -362,13 +362,15 @@ pub trait Approved: Base {
...
         }
 
-        if self.peer_map().has(&their_pub_id) {
-            trace!(
-                "{} - Not sending connection request to {} - already connected.",
-                self,
-                their_pub_id
-            );
-            return Ok(());
+        if let Some(p2p_node) = self.chain().get_p2p_node(their_pub_id.name()) {
```
Comment (madadam) [inline src/states/common/approved.rs]: Why do we need to send the connection request if we have the `P2pNode`?

File: src/states/common/approved.rs [hunk 16]
```diff
@@ -413,13 +415,26 @@ pub trait Approved: Base {
...
-            .insert(their_pub_id, their_conn_info.clone());
+        self.peer_map_mut().connect(their_conn_info.clone());
         self.send_direct_message(&their_conn_info, DirectMessage::ConnectionResponse);
 
         Ok(())
     }
 
+    fn disconnect_by_id_lookup(&mut self, pub_id: &PublicId) {
+        if let Some(node) = self.chain().get_p2p_node(pub_id.name()) {
+            let node = node.clone();
```
Comment (madadam) [inline src/states/common/approved.rs]: Instead of cloning the whole `P2pNode`, you can just copy the peer address.

File: src/states/elder/mod.rs (hunk 12 above)
Comment (octol) [reply src/states/elder/mod.rs]: Removing those `connect` calls does still seem to pass the tests

File: src/states/elder/mod.rs [hunk 17]
```diff
@@ -363,8 +364,7 @@ impl Elder {
 
     fn establish_connection(&mut self, node: P2pNode) {
         self.send_direct_message(node.connection_info(), DirectMessage::ConnectionResponse);
-        self.peer_map_mut()
-            .insert(*node.public_id(), node.into_connection_info());
+        self.peer_map_mut().connect(node.connection_info().clone())
```
Comment (madadam) [inline src/states/elder/mod.rs]: Do we still need to do this? We already have the full connection info. I think it should be sufficient to call `peer_map.connect` only in `handle_connected_to`. This also applies to the other places where we do this. Similarly, we should probably only do `peer_map.disconnect` in `handle_connection_failure`.

File: src/peer_map.rs (hunk 14 above)
Comment (octol) [reply src/peer_map.rs]: This was indeed discussed during planning. The argument for `ConnectionInfo` was that you can return a reference and just use that directly. I think both ways has merit. What do you think @jeanphilippeD ?

Comment (jeanphilippeD) [reply src/peer_map.rs]: We had a chat before, this is simpler and the gain is very small, that avoid having to rebuild the ConnectionInfo everytime we get it.

File: src/states/elder/mod.rs (hunk 17 above)
Comment (octol) [reply src/states/elder/mod.rs]: Agree, I've started removing the `connect` calls to see if it still works

**Comment by jeanphilippeD** [comment]
> Looks mostly OK, just not sure about those explicit peer_map.connect and peer_map.disconnect calls. I don't think we need them anywhere expect handle_connected_to and handle_connection_failure.

@madadam agree with connect. Was not sure for disconnect, do we get an event from quic_p2p when we initiate the disconnect ourselves?

File: src/peer_map.rs (hunk 14 above)
Comment (madadam) [reply src/peer_map.rs]: OK, fair enough. I think at some point in the future we should change the way we store the certificate from `Vec<u8>` to `Rc<[u8]>` (or `Arc<[u8]>`) to make them cheap to clone. Then it wouldn't matter to return them by value instead of reference so we would be able to avoid this duplication.

File: src/states/elder/mod.rs (hunk 17 above)
Comment (octol) [reply src/states/elder/mod.rs]: What about `connect` in `send_bootstrap_request`? I think we still need it there

Comment (madadam) [reply src/states/elder/mod.rs]: I'm not sure. Do we? I think it's the same thing. When we receive a response, that response will be preceded by `ConnectedTo` event from quic-p2p and we would `peer_map.connect` there.

File: src/states/common/approved.rs (hunk 15 above)
Comment (octol) [reply src/states/common/approved.rs]: If we have their `P2pNode` we are connected, but are we also connected at the network layer?

**Comment by madadam** [comment]
> do we get an event from quic_p2p when we initiate the disconnect ourselves?

We don't. We need to disconnect explicitly then. But I think `Base::disconnect` does that already.

File: src/states/common/approved.rs (hunk 15 above)
Comment (madadam) [reply src/states/common/approved.rs]: Not necessarily, but the whole idea behind this work is that it doesn't matter. We make no difference between knowing someones connection info and actually being connected. I think.

File: src/states/elder/mod.rs (hunk 17 above)
Comment (octol) [reply src/states/elder/mod.rs]: All the tests failed when I tried removing that one, just thought I'd check before started investigating the failures

Comment (madadam) [reply src/states/elder/mod.rs]: Hm, OK. Then maybe put it back. There might be something subtle going on in `BootstrappingPeer`. We can always revisit it later.

File: src/states/common/approved.rs (hunk 15 above)
Comment (jeanphilippeD) [reply src/states/common/approved.rs]: (That is once it is finished, but that specific piece of work does not change the model of all node we know are actually connected with quic_p2p).
